python/ray/train/mosaic/mosaic_trainer.py

```diff
@@ -0,0 +1,191 @@
+    Ray Actors. The training is carried out in a distributed fashion through PyTorch
+    DDP. These actors already have the necessary torch process group already
+    configured for distributed PyTorch training.
+
+    The training function ran on every Actor will first run the
+    specified ``trainer_init_per_worker`` function to obtain an instantiated
+    ``composer.Trainer`` object. The ``trainer_init_per_worker`` function
+    will have access to preprocessed train and evaluation datasets.
+
+    Example:
+    >>> def trainer_init_per_worker(config):
+    ...     # prepare the model for distributed training and wrap with
+    ...     # ComposerClassifier for Composer Trainer compatibility
+    ...     model = torchvision.models.resnet18(num_classes=10)
+    ...     model = ComposerClassifier(ray.train.torch.prepare_model(model))
+    ...
+    ...     # prepare train dataloader
+    ...     batch_size_per_worker = BATCH_SIZE // session.get_world_size()
+    ...     train_dataloader = torch.utils.data.DataLoader(
+    ...         config.pop("train_dataset"),
+    ...         batch_size=batch_size_per_worker
+    ...     )
+    ...     train_dataloader = ray.train.torch.prepare_data_loader(train_dataloader)
+    ...
+    ...     # prepare optimizer
+    ...     optimizer = composer.optim.DecoupledSGDW(
+    ...         model.parameters(),
+    ...         lr=0.05,
+    ...         momentum=0.9,
+    ...         weight_decay=2.0e-3,
+    ...     )
+    ...
+    ...     return composer.trainer.Trainer(
+    ...         model=model,
+    ...         train_dataloader=train_dataloader,
+    ...         optimizers=optimizer,
+    ...         **config
+    ...     )
+    ...
+    >>> scaling_config = ScalingConfig(num_workers=2, use_gpu=True)
+    >>> trainer_init_config = {
+    ...     "max_duration": "1ba",
+    ...     "train_dataset": train_dataset,
+    ...     "algorithms": [LabelSmoothing()],
+    ... }
+    ...
+    >>> trainer = MosaicTrainer(
+    ...     trainer_init_per_worker=trainer_init_per_worker,
+    ...     trainer_init_config=trainer_init_config,
+    ...     scaling_config=scaling_config,
+    ... )
+    ...
+    >>> trainer.fit()
+
+
+    Args:
+        trainer_init_per_worker: The function that returns an instantiated
+            ``composer.Trainer`` object and takes in configuration
+            dictionary (``**config``) as an argument. This dictionary is based on
+            ``trainer_init_config`` and is modified for Ray - Composer integration.
+        datasets: Any Ray Datasets to use for training. At the moment, we do not support
+            passing datasets to the trainer and using the dataset shards in the trainer
+            loop. Instead, configure and load the datasets inside
+            ``trainer_init_per_worker`` function
+        trainer_init_config: Configurations to pass into ``trainer_init_per_worker`` as
+            kwargs. To define configurations for the composer trainer's ``fit`` call,
```

Comment:
the question that comes to mind is why do I need this instead of manually specifying these configurations in the `Trainer`.

I think it would be useful to mention in this docstring that `trainer_init_config` is useful for hyperparameter tuning so you can pass in multiple different configurations rather than hard coding it.

Reply:
added further elaboration